src/index.ts

```diff
@@ -77,14 +92,40 @@ export const go = async <T, E extends Error>(
   const fn = typeof predicate === 'function' ? predicate : () => predicate;
   if (!options) return attempt(fn);
 
-  const { retries, timeoutMs } = options;
+  const { retries, timeoutMs, delay, fullTimeoutMs } = options;
+
+  let fullTimeoutExceeded = false;
+  if (fullTimeoutMs !== undefined) {
+    // Start a "full" timeout that will stop all retries after it is exceeded
+    sleep(fullTimeoutMs).then(() => {
+      fullTimeoutExceeded = true;
+    });
+  }
+
   const attempts = retries ? retries + 1 : 1;
   let lastFailedAttemptResult: GoResultError<E> | null = null;
   for (let i = 0; i < attempts; i++) {
+    // This is guaranteed to be false for the first attempt
```

Comment:
How is it guaranteed? If it's set to something very small it can be set to `true` even before the first attempt, can't it?

Reply:
No, because the `sleep` promsie for the global timeout schedules a new microtask and the function code continues to run synchronously reaching this line. That means this will always run before the fullTimeoutExceeded is set to true.